x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,392 @@
+	)
+
+	m.report = report
+
+	endDate := time.Now().UTC()
+
+	if !m.config.Collection.Realtime {
+		// If we're not collecting realtime data, then just pull until
+		// yesterday (in UTC).
+		endDate = endDate.AddDate(0, 0, -1)
+	}
+
+	startDate := endDate.AddDate(0, 0, -m.config.Collection.LookbackDays)
+
+	return m.fetchDateRange(startDate, endDate, httpClient)
+}
+
+// fetchDateRange retrieves OpenAI API usage data for each configured API key within a specified date range.
+//
+// For each API key:
+// 1. Generates a secure SHA-256 hash of the key for state tracking
+// 2. Checks the state store for the last processed date
+// 3. Adjusts start date if previous state exists to avoid duplicate collection
+// 4. Iterates through each day in the range, collecting usage data
+// 5. Updates state store with the latest processed date
+func (m *MetricSet) fetchDateRange(startDate, endDate time.Time, httpClient *RLHTTPClient) error {
+	for _, apiKey := range m.config.APIKeys {
+		// SHA-256 is a cryptographic hash function that generates a 256-bit (32-byte) digest,
+		// represented as a 64-character hexadecimal string. The hash function is deterministic,
+		// ensuring the same input consistently produces identical output, while being computationally
+		// infeasible to reverse. Its strong collision resistance makes it highly suitable for
+		// secure key storage.
+		//
+		// The hexadecimal representation uses only alphanumeric characters [0-9a-f], making it
+		// ideal for cross-platform filename compatibility. The fixed-length output provides
+		// predictable storage requirements and consistent behavior across different systems.
+		hasher := sha256.New()
+		hasher.Write([]byte(apiKey.Key))
+		hashedKey := hex.EncodeToString(hasher.Sum(nil))
+		stateKey := "state_" + hashedKey
+
+		if m.stateStore.Has(stateKey) {
+			lastProcessedDate, err := m.stateStore.Get(stateKey)
+			if err != nil {
+				m.logger.Errorf("Error reading state for API key: %v", err)
+				continue
+			}
+
+			lastDate, err := time.Parse("2006-01-02", lastProcessedDate)
+			if err != nil {
+				m.logger.Errorf("Error parsing last processed date: %v", err)
+				continue
+			}
+
+			startDate = lastDate.AddDate(0, 0, 1)
+			if startDate.After(endDate) {
+				continue
+			}
+		}
+
+		for d := startDate; !d.After(endDate); d = d.AddDate(0, 0, 1) {
+			dateStr := d.Format("2006-01-02")
+			if err := m.fetchSingleDay(dateStr, apiKey.Key, httpClient); err != nil {
+				m.logger.Errorf("Error fetching data for date %s: %v", dateStr, err)
+				continue
+			}
+		}
+
+		if err := m.stateStore.Put(stateKey, endDate.Format("2006-01-02")); err != nil {
+			m.logger.Errorf("Error storing state for API key: %v", err)
+		}
+	}
+	return nil
+}
+
+func (m *MetricSet) fetchSingleDay(dateStr, apiKey string, httpClient *RLHTTPClient) error {
+	req, err := m.createRequest(dateStr, apiKey)
+	if err != nil {
+		return fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, err := httpClient.Do(req)
```

Comment:
[Suggestion/NIT]
as you're not using an interface for the `RLHTTPClient`, why not having `Do` take a context and put here the `context.TODO` or even move it all the way up to `Fetch`?